ntp-proto/src/time_types.rs

```diff
@@ -1,47 +1,50 @@
 use std::ops::{Add, AddAssign, Div, DivAssign, Mul, MulAssign, Sub, SubAssign};
-use std::time::Duration;
+use std::time::{Duration, Instant};
 
 /// NtpInstant is a monotonically increasing value modelling the uptime of the NTP service
 ///
 /// It is used to validate packets that we send out, and to order internal operations.
-#[derive(Debug, Copy, Clone, Eq, PartialEq, PartialOrd, Ord, Default)]
+#[derive(Debug, Copy, Clone, Eq, PartialEq, PartialOrd, Ord)]
 pub struct NtpInstant {
-    timestamp: NtpTimestamp,
+    instant: Instant,
 }
 
 impl NtpInstant {
-    pub const ZERO: Self = Self {
-        timestamp: NtpTimestamp::ZERO,
-    };
-
-    pub const fn from_ntp_timestamp(timestamp: NtpTimestamp) -> Self {
-        Self { timestamp }
-    }
-
-    #[cfg(any(test, feature = "fuzz"))]
-    pub(crate) const fn from_bits(bits: [u8; 8]) -> Self {
+    pub fn now() -> Self {
         Self {
-            timestamp: NtpTimestamp::from_bits(bits),
+            instant: Instant::now(),
         }
     }
 
-    pub(crate) const fn to_bits(self) -> [u8; 8] {
-        self.timestamp.to_bits()
+    // used to populate the T1 and T3 fields to ensure the server is doing something sensible
+    // see `generate_poll_message` for details
+    pub(crate) fn to_bits(self) -> [u8; 8] {
+        use std::collections::hash_map::DefaultHasher;
+        use std::hash::{Hash, Hasher};
+
+        let mut s = DefaultHasher::new();
+        self.instant.hash(&mut s);
+        s.finish().to_be_bytes()
     }
 
-    #[cfg(any(test, feature = "fuzz"))]
-    pub(crate) const fn from_fixed_int(timestamp: u64) -> NtpInstant {
-        Self {
-            timestamp: NtpTimestamp::from_fixed_int(timestamp),
-        }
+    pub fn abs_diff(self, rhs: Self) -> NtpDuration {
+        let duration = if self.instant > rhs.instant {
+            self.instant - rhs.instant
+        } else {
+            rhs.instant - self.instant
+        };
+
+        NtpDuration::from_system_duration(duration)
     }
 }
 
-impl Sub for NtpInstant {
-    type Output = NtpDuration;
+impl Add<Duration> for NtpInstant {
+    type Output = NtpInstant;
 
-    fn sub(self, rhs: Self) -> Self::Output {
-        self.timestamp - rhs.timestamp
+    fn add(mut self, rhs: Duration) -> Self::Output {
+        self.instant += rhs;
+
+        self
     }
 }
```

Comment:
I think we only ever need this in testing code, so we should probably mark it with the appropriate cfg marker.